---
title: "BIS-VIS Methylation Analysis Example" 
author: "Kristy Ou"
date: "4/28/2020"
output: html_document
runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggrepel)
```

BIS-VIS is a tool that analyzes bisulfite amplicon sequencing data, and provides a report of the results containing interactive visuals. 

Bisulfite amplicon seq allows for sensitive detection of DNA methylation at a targeted genomic region. Regions of interest are often selected due to cell-type or disease specific methylation statuses. Raw FASTQ files were processed with [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) and [CutAdapt](https://cutadapt.readthedocs.io/en/stable/). Alignment and methylation calling were performed with [Bismark](https://rawgit.com/FelixKrueger/Bismark/master/Docs/Bismark_User_Guide.html). The resulting methylation coverage files were used as input for BIS-VIS. 

The methylation status of the Regulatory T cell (Treg)-Specific Demethylated Region (**TSDR**) was assessed in several immune cell subsets. The example below showcases a typical report generated by BIS-VIS.

```{r data import, include=FALSE}
## Importing Files 
### Documnent needs to be in the same folder as files!
cov.files <- list.files(pattern = ".cov.gz$")

## Create list of tibbles using purrr
cov.df.list <- cov.files %>% 
  map(function(x){
    read_delim(x, delim = "\t", col_names = c("Chr", "Pos1", "Pos2", "Perc_Meth", "Meth_Reads", "Unmeth_Reads"), col_types = "ciidii")
  })

## Rename list elements using stringr 
names(cov.df.list) <- str_replace(cov.files, pattern = "_R1_bismark_bt2_pe.bismark.cov.gz", replacement = "")

## Filter reads and add sample column
for (i in seq_along(cov.df.list)){
  cov.df.list[[i]] <- cov.df.list[[i]] %>%
    dplyr::filter(Meth_Reads + Unmeth_Reads > 1000) %>%
    select(-Pos2) %>%
    mutate("Sample" = names(cov.df.list[i]))
}

## Final data frame for plotting
meth.data = dplyr::bind_rows(cov.df.list)

## Read and add sample sheet 
sample.sh <- read_csv("sample_sheet.csv")
meth.data <- read_rds("meth.data")

## Add sample features to meth.data
meth.data <- left_join(meth.data, sample.sh, by = "Sample")
```

#### Sample cluster behavior 
```{r PCA, echo = FALSE}
# Create a dataset for PCA analysis
pca.df <- meth.data %>% 
  select(Sample, Pos1, Perc_Meth) %>% 
  pivot_wider(id_cols = Sample, names_from = Pos1, values_from = Perc_Meth) %>% 
  column_to_rownames(var = "Sample")

# Perform PCA analysis
pca.results <- prcomp(pca.df)

pca.variance <- summary(pca.results)[["importance"]]

# Create PCA to plot
pca.plot <- as.data.frame(pca.results$x) %>% 
  rownames_to_column(var = "Sample") %>% 
  left_join(sample.sh, by = "Sample")


### PCA interactive plot
ui <- fluidPage(

    sidebarLayout(
        sidebarPanel(
            selectInput(inputId = "Color.By", label = "Select color code", 
                        choices = str_subset(colnames(pca.plot),"PC", negate = TRUE)), 
            
            selectInput(inputId = "pc.x", label = "Select which PC to view on X axis",
                        choices = str_subset(colnames(pca.plot),"PC"))
            
            ),
        
        mainPanel(
           plotOutput(outputId = "pca")
        )
    )
)

# Server logic ---- 
server <- function(input, output){
    output$pca <- renderPlot({
        ggplot(pca.plot, aes_string(input$pc.x, colnames(pca.plot)[str_which(colnames(pca.plot),input$pc.x)+1]))+
          geom_jitter(aes_string(color = input$Color.By))+
          geom_text_repel(aes(label = Sample)) +
          theme_bw()
    })
}

# Run app ----
shinyApp(ui = ui, server = server)
```


#### TSDR Methylation by Samples 
```{r Meth by samples, echo=FALSE}
ui <- fluidPage(

    sidebarLayout(
        sidebarPanel(
            # Select which samples to analyze
            checkboxGroupInput(inputId = "Select.Samples", label = "Select individual samples to view", 
                               choices = unique(meth.data$Sample))
            ),
        
        mainPanel(
           plotOutput(outputId = "scatterplot1")
        )
    )
)

# Server logic ---- 
server <- function(input, output){

    output$scatterplot1 <- renderPlot({
        ggplot(data = subset(meth.data, Sample %in% input$Select.Samples), aes(Pos1, Perc_Meth))+
            geom_line(aes(color=Sample), alpha = 0.5)+
            geom_point(aes(fill=Sample), shape = 23, color = "black", size = 3.5)+
            theme_bw()+
            labs(x="Chromosomal Position", y="Methylation %")+
            ylim(0,100)+
            scale_x_continuous(breaks = seq.int(49260550, 49260950, 50))
    })
}

# Run app ----
shinyApp(ui = ui, server = server)
```

#### TSDR Methylation by Group
```{r Meth by group, echo=F}
ui <- fluidPage(
    
  sidebarLayout(
        sidebarPanel(
            selectInput(inputId = "Select.Group", label = "Select which group to analyze", 
                        choices = str_subset(colnames(pca.plot),"PC|Sample|ID", negate = TRUE))
            ),
        
        mainPanel(
           plotOutput(outputId = "group.plot")
        )
    )
)
  
# Server logic ---- 
server <- function(input, output){

    output$group.plot <- renderPlot({
      ggplot(meth.data, aes(Pos1, Perc_Meth))+
        stat_summary(aes_string(group=input$Select.Group, color = input$Select.Group), fun.y=mean, geom="line")+
        geom_point(aes_string(fill=input$Select.Group), shape = 23, color = "black", size = 2, alpha = 0.3)+
        theme_bw()+
        labs(x="Chromosomal Position", y="Methylation %")+
        ylim(0,100)+
        facet_wrap(~Cell_Type)
    })
}

# Run app ----
shinyApp(ui = ui, server = server)
```











